File: frontend/lib/artifact-semantic.ts
import type { Artifact, ArtifactSearchMatch } from "./artifacts.ts";
import { exactKeys, finiteInteger, objectValue, sameUuid } from "./wire-guards.ts";

const digest = (value: unknown): value is string => typeof value === "string" && /^[0-9a-f]{64}$/.test(value);
const label = (value: unknown, maximum: number): value is string => typeof value === "string" && Array.from(value).length > 0 && Array.from(value).length <= maximum;
export type ArtifactPassage = {
  passage_id: string; artifact_revision: number; text_sha256: string;
  start_offset: number; end_offset: number; model: string; chunk_config: string;
  token_count: number; token_limit: number; cosine_similarity: number;
  score_is_probability: false; score_type: "cosine_similarity";
};
export type ArtifactEmbeddingCoverage = {
  model: string; chunk_config: string; state: "ready" | "incomplete" | "unavailable";
  ready: number; pending: number; processing: number; failed: number; unavailable: number;
  withheld: number; empty: number; passages: number; truncated: number;
  total_kind: "ranked_candidates"; score_type: "semantic_reciprocal_rank";
};
const counters = ["ready", "pending", "processing", "failed", "unavailable", "withheld", "empty", "passages", "truncated"] as const;
export function decodeEmbeddingCoverage(value: unknown): ArtifactEmbeddingCoverage {
  const row = objectValue(value);
  if (!row || !exactKeys(row, ["model", "chunk_config", "state", ...counters, "total_kind", "score_type"])
    || !label(row.model, 300) || !label(row.chunk_config, 100) || !counters.every((field) => finiteInteger(row[field]))
    || row.total_kind !== "ranked_candidates" || row.score_type !== "semantic_reciprocal_rank") throw new Error("Mnemonic returned invalid artifact embedding coverage.");
  const coverage = row as ArtifactEmbeddingCoverage;
  const partial = coverage.pending + coverage.processing + coverage.failed + coverage.unavailable + coverage.withheld + coverage.truncated > 0;
  const state = coverage.ready === 0 && coverage.unavailable > 0 ? "unavailable" : partial ? "incomplete" : "ready";
  if (coverage.state !== state || coverage.truncated > coverage.ready + coverage.pending + coverage.processing + coverage.failed + coverage.unavailable + coverage.empty || coverage.passages < coverage.ready || coverage.ready === 0 && coverage.passages !== 0) throw new Error("Mnemonic returned inconsistent artifact embedding coverage.");
  return coverage;
}
export function decodeArtifactPassage(value: unknown, artifact: Artifact, coverage: ArtifactEmbeddingCoverage, snippet: string | null): ArtifactPassage {
  const row = objectValue(value);
  if (!row || !exactKeys(row, ["passage_id", "artifact_revision", "text_sha256", "start_offset", "end_offset", "model", "chunk_config", "token_count", "token_limit", "cosine_similarity", "score_is_probability", "score_type"])
    || !digest(row.passage_id) || !digest(row.text_sha256) || row.artifact_revision !== artifact.revision
    || artifact.deleted_at !== null || !artifact.content_available || artifact.extraction.status !== "ready"
    || !finiteInteger(row.start_offset, 0, 8_000_000) || !finiteInteger(row.end_offset, 1, 8_000_000)
    || row.end_offset <= row.start_offset || row.end_offset - row.start_offset > 1500
    || row.model !== coverage.model || row.chunk_config !== coverage.chunk_config
    || !finiteInteger(row.token_count, 1, 8192) || !finiteInteger(row.token_limit, 1, 8192) || row.token_count > row.token_limit
    || typeof row.cosine_similarity !== "number" || !Number.isFinite(row.cosine_similarity) || Math.abs(row.cosine_similarity) > 1
    || row.score_is_probability !== false || row.score_type !== "cosine_similarity"
    || typeof snippet !== "string" || Array.from(snippet).length !== Math.min(1000, row.end_offset - row.start_offset)) throw new Error("Mnemonic returned invalid artifact passage evidence.");
  return row as ArtifactPassage;
}
export async function verifyArtifactPassageIds(items: ArtifactSearchMatch[]): Promise<void> {
  for (const { artifact, passage } of items) {
    if (!passage) continue;
    const key = JSON.stringify([artifact.id.toLowerCase(), passage.artifact_revision, passage.text_sha256, passage.model, passage.chunk_config, passage.start_offset, passage.end_offset])
      .replace(/[\x7f-\uffff]/g, (character) => `\\u${character.charCodeAt(0).toString(16).padStart(4, "0")}`);
    const bytes = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(key));
    const identity = Array.from(new Uint8Array(bytes), (byte) => byte.toString(16).padStart(2, "0")).join("");
    if (identity !== passage.passage_id) throw new Error("Mnemonic returned passage evidence for a different source.");
  }
}
export function artifactPassageQuery(passage: ArtifactPassage): URLSearchParams {
  return new URLSearchParams({ expected_revision: String(passage.artifact_revision), expected_text_sha256: passage.text_sha256,
    offset: String(passage.start_offset), limit: String(passage.end_offset - passage.start_offset) });
}
export function decodeArtifactPassageText(value: unknown, match: ArtifactSearchMatch): string {
  const row = objectValue(value), extraction = objectValue(row?.extraction), passage = match.passage;
  if (!passage || !row || !exactKeys(row, ["project_id", "artifact_id", "revision", "sha256", "text_sha256", "extraction", "text", "offset", "limit", "total_chars", "next_offset"])
    || !sameUuid(row.project_id, match.artifact.project_id) || !sameUuid(row.artifact_id, match.artifact.id)
    || row.revision !== passage.artifact_revision || row.sha256 !== match.artifact.sha256 || row.text_sha256 !== passage.text_sha256
    || !extraction || !exactKeys(extraction, ["status", "truncated", "error_code", "extracted_at"])
    || extraction.status !== "ready" || extraction.truncated !== match.artifact.extraction.truncated || extraction.error_code !== null
    || !(extraction.extracted_at === null || typeof extraction.extracted_at === "string" && extraction.extracted_at.length <= 40 && Number.isFinite(Date.parse(extraction.extracted_at)))
    || row.offset !== passage.start_offset || row.limit !== passage.end_offset - passage.start_offset
    || !finiteInteger(row.total_chars, passage.end_offset, 8_000_000)
    || row.next_offset !== (passage.end_offset < row.total_chars ? passage.end_offset : null)
    || typeof row.text !== "string" || Array.from(row.text).length !== row.limit
    || Array.from(row.text).slice(0, 1000).join("") !== match.snippet) throw new Error("Mnemonic returned artifact text outside the selected passage.");
  return row.text;
}

File: mcp/src/mnemonic_mcp/artifact_semantic.py
"""Bounded semantic passages and honest embedding coverage at the agent boundary."""

import hashlib
import json
from typing import Annotated, Literal, Self
from uuid import UUID

from pydantic import BaseModel, BeforeValidator, ConfigDict, Field, StrictInt, model_validator
from pydantic_core import PydanticCustomError

from .input_errors import InputValidationError
from .search_query import QueryMode, constrained_query

Digest = Annotated[str, Field(pattern=r"^[0-9a-f]{64}$")]
Count = Annotated[StrictInt, Field(ge=0)]
EMBEDDING_COUNTS = ("ready", "pending", "processing", "failed", "unavailable", "withheld",
                    "empty", "passages", "truncated")


def validate_artifact_semantic(query: str | None, mode: QueryMode, fulltext: bool,
                               selected: bool = True) -> None:
    if constrained_query(query, mode):
        raise PydanticCustomError("semantic_requires_unconstrained_artifact_query",
                                  "Semantic artifact search requires unconstrained terms")
    if not fulltext:
        raise PydanticCustomError("artifact_semantic_requires_fulltext",
                                  "Artifact semantic search requires fulltext")
    if not (query or "").strip() or not selected:
        raise PydanticCustomError("artifact_semantic_requires_query_and_facet",
                                  "Artifact semantic search requires query and facet")


def _false_probability(value: object) -> Literal[False]:
    if value is not False:
        raise ValueError("Semantic scores are not probabilities")
    return False


class ArtifactPassageEvidence(BaseModel):
    model_config = ConfigDict(extra="forbid")
    passage_id: Digest
    artifact_revision: Annotated[StrictInt, Field(ge=1)]
    text_sha256: Digest
    start_offset: Annotated[StrictInt, Field(ge=0, le=8_000_000)]
    end_offset: Annotated[StrictInt, Field(ge=1, le=8_000_000)]
    model: Annotated[str, Field(min_length=1, max_length=300)]
    chunk_config: Annotated[str, Field(min_length=1, max_length=100)]
    token_count: Annotated[StrictInt, Field(ge=1, le=8192)]
    token_limit: Annotated[StrictInt, Field(ge=1, le=8192)]
    cosine_similarity: Annotated[float, Field(ge=-1, le=1, allow_inf_nan=False, strict=True)]
    score_is_probability: Annotated[Literal[False], BeforeValidator(_false_probability)] = False
    score_type: Literal["cosine_similarity"] = "cosine_similarity"

    @model_validator(mode="after")
    def bounded_span(self) -> Self:
        if not 0 < self.end_offset - self.start_offset <= 1500 or (
            self.token_count > self.token_limit or self.score_is_probability
        ):
            raise ValueError("Passage evidence must describe a bounded, non-probabilistic span")
        return self

    def matches_identity(self, identity: UUID, revision: int) -> bool:
        expected = hashlib.sha256(json.dumps([
            str(identity), revision, self.text_sha256, self.model, self.chunk_config,
            self.start_offset, self.end_offset,
        ], separators=(",", ":")).encode()).hexdigest()
        return self.artifact_revision == revision and self.passage_id == expected


class ArtifactEmbeddingCoverage(BaseModel):
    model_config = ConfigDict(extra="forbid")
    model: Annotated[str, Field(min_length=1, max_length=300)]
    chunk_config: Annotated[str, Field(min_length=1, max_length=100)]
    state: Literal["ready", "incomplete", "unavailable"]
    ready: Count = 0
    pending: Count = 0
    processing: Count = 0
    failed: Count = 0
    unavailable: Count = 0
    withheld: Count = 0
    empty: Count = 0
    passages: Count = 0
    truncated: Count = 0
    total_kind: Literal["ranked_candidates"] = "ranked_candidates"
    score_type: Literal["semantic_reciprocal_rank"] = "semantic_reciprocal_rank"

    @property
    def partial_vectors(self) -> bool:
        return bool(self.pending or self.processing or self.failed or self.unavailable)

    @model_validator(mode="after")
    def coherent_coverage(self) -> Self:
        incomplete = self.partial_vectors or self.withheld or self.truncated
        expected = "unavailable" if not self.ready and self.unavailable else (
            "incomplete" if incomplete else "ready")
        available = sum(getattr(self, name) for name in (
            "ready", "pending", "processing", "failed", "unavailable", "empty"))
        if (self.state != expected or self.passages < self.ready or self.truncated > available
                or not self.ready and self.passages):
            raise ValueError("Embedding coverage counts disagree with its state")
        return self


class ArtifactMatchEvidence(BaseModel):
    model_config = ConfigDict(extra="forbid")
    evidence: Literal["lexical", "semantic"] = "lexical"
    passage: ArtifactPassageEvidence | None = None

    @model_validator(mode="after")
    def coherent_evidence(self) -> Self:
        if (self.evidence == "semantic") != (self.passage is not None):
            raise ValueError("Only semantic artifact evidence carries a passage")
        return self


def artifact_evidence_matches(item, semantic: bool,
                              embedding: ArtifactEmbeddingCoverage | None = None) -> bool:
    if item.evidence != ("semantic" if semantic else "lexical"):
        return False
    if not semantic:
        return item.passage is None
    passage, artifact = item.passage, item.artifact
    if passage is None or embedding is None:
        return False
    return (item.matched_fields == ["content"] and artifact.deleted_at is None
            and artifact.content_available and artifact.extraction.status == "ready"
            and passage.matches_identity(artifact.id, artifact.revision)
            and (passage.model, passage.chunk_config) == (embedding.model, embedding.chunk_config)
            and item.snippet is not None
            and len(item.snippet) == min(1000, passage.end_offset - passage.start_offset))


def validate_tool_artifact_semantic(query: str | None, mode: QueryMode, fulltext: bool) -> None:
    from .validation_rules import VALIDATION_RULES

    try:
        validate_artifact_semantic(query, mode, fulltext)
    except PydanticCustomError as error:
        field, message = VALIDATION_RULES[error.type]
        raise InputValidationError(f"Mnemonic rejected the input. Check: {field} "
                        f"({error.type}). {message}") from None
